File: src/guandan/cards.py
from enum import IntEnum
from typing import NamedTuple


class Suit(IntEnum):
    SPADE = 0
    HEART = 1
    DIAMOND = 2
    CLUB = 3


class Rank(IntEnum):
    TWO = 2
    THREE = 3
    FOUR = 4
    FIVE = 5
    SIX = 6
    SEVEN = 7
    EIGHT = 8
    NINE = 9
    TEN = 10
    JACK = 11
    QUEEN = 12
    KING = 13
    ACE = 14
    # 15 is reserved for level_order_key
    BLACK_JOKER = 16
    RED_JOKER = 17


class Card(NamedTuple):
    rank: int
    suit: int   # 0-3 for normal cards, 0/1 for jokers
    deck: int   # 0 or 1 (which copy from the double deck)


class ComboType(IntEnum):
    PASS = 0
    # Ordinary (7 types)
    SINGLE = 1
    PAIR = 2
    TRIPLE = 3
    FULL_HOUSE = 4      # triple + pair (5 cards)
    STRAIGHT = 5        # exactly 5 consecutive, natural order, NOT all same suit
    TUBE = 6            # exactly 3 consecutive pairs (6 cards), natural order
    PLATE = 7           # exactly 2 consecutive triples (6 cards), natural order
    # Bombs (9 tiers, ordered low→high by enum value)
    BOMB_4 = 8          # quadruple
    BOMB_5 = 9          # quintuple
    STRAIGHT_FLUSH = 10 # 5 consecutive same suit (between 5-of-a-kind and 6-of-a-kind)
    BOMB_6 = 11         # sextuple
    BOMB_7 = 12         # septuple
    BOMB_8 = 13         # octuple
    BOMB_9 = 14         # nonuple (requires wilds)
    BOMB_10 = 15        # decuple (requires wilds)
    BOMB_JOKER = 16     # 2BJ + 2RJ — highest bomb


BOMB_TYPES = frozenset({
    ComboType.BOMB_4, ComboType.BOMB_5, ComboType.STRAIGHT_FLUSH,
    ComboType.BOMB_6, ComboType.BOMB_7, ComboType.BOMB_8,
    ComboType.BOMB_9, ComboType.BOMB_10, ComboType.BOMB_JOKER,
})

# Map N-of-a-kind size to bomb ComboType
BOMB_SIZE_TO_TYPE: dict[int, ComboType] = {
    4: ComboType.BOMB_4, 5: ComboType.BOMB_5,
    6: ComboType.BOMB_6, 7: ComboType.BOMB_7,
    8: ComboType.BOMB_8, 9: ComboType.BOMB_9,
    10: ComboType.BOMB_10,
}


def level_order_key(rank: int, level_rank: int) -> int:
    """Comparison key for LEVEL ORDER.

    Level cards rank above A(14), below BJ(16). Returns 15 for level cards.
    Used for: singles, pairs, triples, full houses, N-of-a-kind bombs.
    """
    if rank == level_rank:
        return 15  # above A(14), below BJ(16)
    return rank


def make_deck() -> list[Card]:
    """Create the 108-card double deck."""
    cards: list[Card] = []
    for deck_id in range(2):
        for rank in range(2, 15):  # 2 through A(=14)
            for suit in range(4):
                cards.append(Card(rank, suit, deck_id))
        cards.append(Card(Rank.BLACK_JOKER, 0, deck_id))
        cards.append(Card(Rank.RED_JOKER, 1, deck_id))
    return cards


def is_wild(card: Card, level_rank: int) -> bool:
    """Is this card a wild (♥ of level rank)?"""
    return card.rank == level_rank and card.suit == Suit.HEART

File: ml/src/guandan_rs/src/cards.rs
// Suit constants (matching Python IntEnum)
pub const SUIT_SPADE: u8 = 0;
pub const SUIT_HEART: u8 = 1;
pub const SUIT_DIAMOND: u8 = 2;
pub const SUIT_CLUB: u8 = 3;

// Rank constants (matching Python IntEnum)
pub const RANK_TWO: u8 = 2;
pub const RANK_THREE: u8 = 3;
pub const RANK_FOUR: u8 = 4;
pub const RANK_FIVE: u8 = 5;
pub const RANK_SIX: u8 = 6;
pub const RANK_SEVEN: u8 = 7;
pub const RANK_EIGHT: u8 = 8;
pub const RANK_NINE: u8 = 9;
pub const RANK_TEN: u8 = 10;
pub const RANK_JACK: u8 = 11;
pub const RANK_QUEEN: u8 = 12;
pub const RANK_KING: u8 = 13;
pub const RANK_ACE: u8 = 14;
// 15 is reserved for level_order_key
pub const RANK_BLACK_JOKER: u8 = 16;
pub const RANK_RED_JOKER: u8 = 17;

// ComboType constants (matching Python IntEnum)
pub const PASS: u8 = 0;
pub const SINGLE: u8 = 1;
pub const PAIR: u8 = 2;
pub const TRIPLE: u8 = 3;
pub const FULL_HOUSE: u8 = 4;
pub const STRAIGHT: u8 = 5;
pub const TUBE: u8 = 6;
pub const PLATE: u8 = 7;
pub const BOMB_4: u8 = 8;
pub const BOMB_5: u8 = 9;
pub const STRAIGHT_FLUSH: u8 = 10;
pub const BOMB_6: u8 = 11;
pub const BOMB_7: u8 = 12;
pub const BOMB_8: u8 = 13;
pub const BOMB_9: u8 = 14;
pub const BOMB_10: u8 = 15;
pub const BOMB_JOKER: u8 = 16;

#[derive(Clone, Copy, Hash, Eq, PartialEq, Debug)]
pub struct Card {
    pub rank: u8,
    pub suit: u8,
    pub deck: u8,
}

impl Card {
    pub fn new(rank: u8, suit: u8, deck: u8) -> Self {
        Card { rank, suit, deck }
    }
}

pub fn level_order_key(rank: u8, level_rank: u8) -> u8 {
    if rank == level_rank {
        15
    } else {
        rank
    }
}

pub fn is_wild(card: &Card, level_rank: u8) -> bool {
    card.rank == level_rank && card.suit == SUIT_HEART
}

pub fn is_bomb_type(combo_type: u8) -> bool {
    matches!(
        combo_type,
        BOMB_4 | BOMB_5 | STRAIGHT_FLUSH | BOMB_6 | BOMB_7 | BOMB_8 | BOMB_9 | BOMB_10
            | BOMB_JOKER
    )
}

pub fn bomb_size_to_type(size: usize) -> Option<u8> {
    match size {
        4 => Some(BOMB_4),
        5 => Some(BOMB_5),
        6 => Some(BOMB_6),
        7 => Some(BOMB_7),
        8 => Some(BOMB_8),
        9 => Some(BOMB_9),
        10 => Some(BOMB_10),
        _ => None,
    }
}

pub fn make_deck() -> Vec<Card> {
    let mut cards = Vec::with_capacity(108);
    for deck_id in 0..2u8 {
        for rank in 2..=14u8 {
            for suit in 0..4u8 {
                cards.push(Card::new(rank, suit, deck_id));
            }
        }
        cards.push(Card::new(RANK_BLACK_JOKER, 0, deck_id));
        cards.push(Card::new(RANK_RED_JOKER, 1, deck_id));
    }
    cards
}

/// Stable id in [0, 108) matching Python's `card_to_id`.
///
/// Layout:
/// - ranks 2..=A × 4 suits × 2 decks = 104 ids (0..=103)
///   - id = (rank − 2) × 8 + suit × 2 + deck
/// - BLACK_JOKER deck 0 → 104, deck 1 → 105
/// - RED_JOKER   deck 0 → 106, deck 1 → 107
pub fn card_to_id(card: &Card) -> usize {
    if card.rank == RANK_BLACK_JOKER {
        return 104 + card.deck as usize;
    }
    if card.rank == RANK_RED_JOKER {
        return 106 + card.deck as usize;
    }
    (card.rank as usize - 2) * 8 + card.suit as usize * 2 + card.deck as usize
}
